wagtail/models/__init__.py

```diff
@@ -864,6 +864,60 @@ def get_lock(self):
 class WorkflowMixin:
     """A mixin that allows a model to have workflows."""
 
+    @classmethod
+    def check(cls, **kwargs):
+        return [
+            *super().check(**kwargs),
+            *cls._check_draftstate_and_revision_mixins(),
+            *cls._check_lockable_mixin(),
+        ]
+
+    @classmethod
+    def _check_draftstate_and_revision_mixins(cls):
+        mro = cls.mro()
+        error = checks.Error(
+            "WorkflowMixin requires DraftStateMixin and RevisionMixin "
+            "(in that order).",
+            hint=(
+                "Make sure your model's inheritance order is as follows: "
+                "WorkflowMixin, DraftStateMixin, and RevisionMixin."
+            ),
+            obj=cls,
+            id="wagtailcore.E006",
+        )
+
+        try:
+            if not (
+                mro.index(WorkflowMixin)
+                < mro.index(DraftStateMixin)
+                < mro.index(RevisionMixin)
+            ):
+                return [error]
+        except ValueError:
+            return [error]
+
+        return []
+
+    @classmethod
+    def _check_lockable_mixin(cls):
+        mro = cls.mro()
+        error = checks.Error(
+            "WorkflowMixin requires LockableMixin to be applied after WorkflowMixin.",
+            hint="Add LockableMixin to the model's base classes after WorkflowMixin.",
+            obj=cls,
+            id="wagtailcore.E007",
+        )
+        try:
+            # WorkflowMixin calls super().get_lock() in the override,
+            # so if applied, it must be applied before LockableMixin.
+            if mro.index(LockableMixin) < mro.index(WorkflowMixin):
```

Comment:
Hmm - I think our usages of "applied before" and "after" are at odds here! I'd think "applied before" meant "applied closer to the base class", ie to the right, and "applied after" would mean it was closer to being the final layer in the inheritance, so closer to the left. That said, our other errors go with this wording, so maybe it's just me who reads it that way! Definitely not a blocker, but maybe something for us to look at rewording how we write inheritance errors if anyone else turns out to find this confusing.

Reply:
Yeah this is tricky, I initially had the same thought as yours when I wrote the first check for `DraftStateMixin`. The thing is, you can say it can be confusing if we go that way as well: "you say it needs to be *applied before*, but why do you actually need to *write it after*?"

I decided following the MRO for the "before/after" makes more sense here, because it's consistent:
- If you write `WorkflowMixin, LockableMixin` let's state this as applying `WorkflowMixin` *before* `LockableMixin`
- Methods from `WorkflowMixin` will be called *before* those from `LockableMixin`
- When you provide the hint to "put `WorkflowMixin` before `LockableMixin`", it's exactly what it reads: `WorkflowMixin, LockableMixin`.

Reply:
Yeah, I suspect it'll be obvious in context anyway - if it's complaining about the model inheritance being one way round, it's probably natural to try the other way